rollup/rollupsyncservice/rollupsyncservice.go

```diff
@@ -0,0 +1,395 @@
+func (s *RollupSyncService) decodeChunkBlockRanges(txData []byte) ([]*rawdb.ChunkBlockRange, error) {
+	const methodIDLength = 4
+	if len(txData) < methodIDLength {
+		return nil, fmt.Errorf("transaction data is too short")
+	}
+
+	method, err := s.scrollChainABI.MethodById(txData[:methodIDLength])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get method by ID, ID: %v, err: %w", txData[:4], err)
+	}
+
+	values, err := method.Inputs.Unpack(txData[methodIDLength:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack transaction data using ABI: %v", err)
+	}
+
+	type commitBatchArgs struct {
+		Version                uint8
+		ParentBatchHeader      []byte
+		Chunks                 [][]byte
+		SkippedL1MessageBitmap []byte
+	}
+	var args commitBatchArgs
+	err = method.Inputs.Copy(&args, values)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode calldata into commitBatch args, err: %w", err)
+	}
+
+	if args.Version != batchHeaderVersion {
+		return nil, fmt.Errorf("unexpected batch version, expected: %d, got: %v", batchHeaderVersion, args.Version)
+	}
+
+	return DecodeChunkBlockRanges(args.Chunks)
+}
+
+func calculateFinalizedBatchMeta(parentBatchMeta *rawdb.FinalizedBatchMeta, batchHash common.Hash, chunks []*Chunk) rawdb.FinalizedBatchMeta {
+	totalL1MessagePopped := parentBatchMeta.TotalL1MessagePopped
+	for _, chunk := range chunks {
+		totalL1MessagePopped += chunk.NumL1Messages(totalL1MessagePopped)
+	}
+	return rawdb.FinalizedBatchMeta{
+		BatchHash:            batchHash,
+		TotalL1MessagePopped: totalL1MessagePopped,
+	}
+}
+
+// validateBatch verifies the consistency between l1 contract and l2 node data.
+// close the node and exit once any consistency check fails.
+func validateBatch(event *L1FinalizeBatchEvent, parentBatchMeta *rawdb.FinalizedBatchMeta, chunks []*Chunk, node *node.Node) error {
+	if len(chunks) == 0 {
+		return fmt.Errorf("invalid arg: length of chunks is 0")
+	}
+	endChunk := chunks[len(chunks)-1]
+	if len(endChunk.Blocks) == 0 {
+		return fmt.Errorf("invalid arg: block number of last chunk is 0")
+	}
+	endBlock := endChunk.Blocks[len(endChunk.Blocks)-1]
+	localWithdrawRoot := endBlock.WithdrawRoot
+	if localWithdrawRoot != event.WithdrawRoot {
+		log.Error("Withdraw root mismatch", "l1 withdraw root", event.WithdrawRoot.Hex(), "l2 withdraw root", localWithdrawRoot.Hex())
+		syscall.Kill(os.Getpid(), syscall.SIGTERM)
+		return fmt.Errorf("Withdraw root mismatch")
+	}
+
+	localStateRoot := endBlock.Header.Root
+	if localStateRoot != event.StateRoot {
+		log.Error("State root mismatch", "l1 state root", event.StateRoot.Hex(), "l2 state root", localStateRoot.Hex())
+		syscall.Kill(os.Getpid(), syscall.SIGTERM)
+		return fmt.Errorf("State root mismatch")
+	}
+
+	batchHeader, err := NewBatchHeader(batchHeaderVersion, event.BatchIndex.Uint64(), parentBatchMeta.TotalL1MessagePopped, parentBatchMeta.BatchHash, chunks)
```

Comment:
At first I was worried that this PR uses `TotalL1MessagePopped` fetched from L1. But it seems like we're calculating it locally, so it should be fine.

Btw might be worth adding a comment explaining that if the batch headers match, then the blocks and transactions (including skipping) must also match.

Reply:
Yes. It's totally calculated locally.

Reply:
two comments are added in [41f07c7](https://github.com/scroll-tech/go-ethereum/pull/515/commits/41f07c7245a2cfc8e9e42aad247e746554c541b6)